File: two-sum/go/medium_optimized.go
package main

func TwoNumberSum(array []int, target int) []int {	
	var numS = make(map[int]bool)
	for _, num := range array {
		tempTarget := target-num
		if _, ok := numS[tempTarget]; ok {
			return []int{tempTarget, num}
		}
		numS[num]  = true
	}
	return []int{}
}

File: two-sum/java/medium_optimized.java
import java.util.*;

class Program {
  public static int[] twoNumberSum(int[] array, int targetSum) {
    // Write your code here.
    Map<Integer, Boolean> store = new HashMap<>();
    for(int num : array){
      int tempTarget = targetSum - num;
      Boolean check = store.get(tempTarget);
      if(check!=null && check){
        return new int[]{tempTarget, num};
      }
      store.put(num, true);
    }
    return new int[0];
  }
}
